pkg/controller/statusmanager/status_manager.go

```diff
@@ -49,8 +53,48 @@ type StatusManager struct {
 	relatedObjects []configv1.ObjectReference
 }
 
-func New(client client.Client, name, version string) *StatusManager {
-	return &StatusManager{client: client, name: name, version: version}
+func New(client client.Client, mapper meta.RESTMapper, name, version string) *StatusManager {
+	return &StatusManager{client: client, mapper: mapper, name: name, version: version}
+}
+
+// deleteRelatedObjects checks for related objects attached to ClusterOperator and deletes
+// whatever is not been rendered from manifests. This is a mechanism to cleanup objects
+// that are no longer needed and are probably present from a previous version
+func (status *StatusManager) deleteRelatedObjectsNotRendered(co *configv1.ClusterOperator) {
+	if status.relatedObjects != nil {
+		for _, current_obj := range co.Status.RelatedObjects {
+			var found bool = false
+			for _, rendered_obj := range status.relatedObjects {
+				found = reflect.DeepEqual(current_obj, rendered_obj)
+
+				if found {
+					break
+				}
+			}
+			if !found {
+				gvr := schema.GroupVersionResource{
+					Group:    current_obj.Group,
+					Resource: current_obj.Resource,
+				}
+				gvk, err := status.mapper.KindFor(gvr)
+				log.Printf("Detected related object with GVK %+v, namespace %v and name %v not rendered by manifests, deleting...", gvk, current_obj.Namespace, current_obj.Name)
+				if err != nil {
+					log.Printf("Error getting GVK of object for deletion: %v", err)
```

Comment:
In this case, should we be re-adding this to relatedObjects? Probably.

Reply:
Not sure I follow.
If there's an error, the status.relatedObjects never changes, if we re-add the object wouldn't we have duplicates?

Reply:
Right - my read of the code is that this would be equivalent to a failure to delete. This means that we've now lost the reference to this object, right?

To be fair, this is very unlikely, so I don't think we really need to worry about it.

Reply:
Yeah, but what relatedObjects hold are not really references to objects. They are instances of configv1 ObjectReference, which are merely a struct containing Group/Resource/Name/Namespace.
So even if we had a failure on delete on the real object, we would still have a valid relatedObject for it, which is just a representation, not a reference perse.